src/middlewares.ts

```diff
@@ -15,37 +15,85 @@ import {
 	touchCommit,
 } from './api';
 
-export async function determineCommitHash(req: any, res: any, next: any) {
-	const isHashInSession = !!req.session.commitHash;
-	const isHashSpecified = req.query && (req.query.hash || req.query.branch);
-	let commitHash;
+function assembleSubdomainUrlForHash(req: any, commitHash: string) {
+    const protocol = (req.secure ? 'https' : 'http');
 
-	if (isHashInSession && !isHashSpecified) {
-		next();
-		return;
-	}
+    return protocol + '://hash-' + commitHash + '.' + stripCommitHashSubdomainFromHost(req.headers.host);
+}
 
-	if (!isHashSpecified) {
-		res.send('Please specify a branch to load');
-		return;
-	}
+function stripCommitHashSubdomainFromHost(host: string) {
+    let segments = host.split('.'),
+        commitHashIndex = null;
+
+    for (let i = 0; i < segments.length; i += 1) {
+        if (0 === segments[i].indexOf('hash-')) {
+            commitHashIndex = i;
+            break;
+        }
+    }
+
+    if (null === commitHashIndex) {
+        return host;
+    }
+
+    return segments.slice(commitHashIndex + 1).join('.');
+}
+
+function getCommitHashFromSubdomain(req: any) {
+    const commitHash = _.find(
+        req.headers.host.split('.'),
+        function (hostSegment) {
+            return 0 === hostSegment.indexOf('hash-');
+        }
+    );
+
+    return commitHash.replace(/^hash-/, '');
+}
 
-	if (req.query.hash) {
+export async function redirectHashFromQueryStringToSubdomain(req: any, res: any, next: any) {
+    const isHashSpecified = req.query && (req.query.hash || req.query.branch);
+
+    let commitHash;
+
+    if (!isHashSpecified) {
+        next();
+        return;
+    }
+
+    if (req.query.hash) {
 		commitHash = req.query.hash;
 	} else if (req.query.branch) {
 		commitHash = getCommitHashForBranch(req.query.branch);
-	}
+    }
 
-	if (commitHash instanceof Error) {
-		res.send(striptags('Calypso Server: ' + commitHash.message));
+    res.writeHead(
```

Comment:
can probably just use response.redirect here https://expressjs.com/en/api.html#res.redirect

Reply:
Thanks for the pointer.  I think I was looking at core Node HTTP docs rather than Express.